model/detector_helpers.py

```diff
@@ -286,3 +295,112 @@ def set_detector_distance(detector, distance):
     fast_axis = detector[0].get_fast_axis()
     slow_axis = detector[0].get_slow_axis()
     detector[0].set_frame(fast_axis, slow_axis, origin)
+
+
+def project_2d(detector):
+    """
+    Project panel origin, fast and slow onto the best-fitting 2D plane.
+    """
+
+    # Extract panel vertices
+    vertices = []
+    for panel in detector:
+        origin = matrix.col(panel.get_origin())
+        fast = matrix.col(panel.get_fast_axis())
+        slow = matrix.col(panel.get_slow_axis())
+        panel_size = panel.get_image_size_mm()
+        point1 = origin + panel_size[0] * fast
+        point2 = origin + panel_size[1] * slow
+        point3 = origin + panel_size[0] * fast + panel_size[1] * slow
+        vertices.append(origin)
+        vertices.append(point1)
+        vertices.append(point2)
+        vertices.append(point3)
+
+    # Fit a plane by SVD. Modified from https://stackoverflow.com/a/18968498
+    points = np.array(vertices).transpose()
+    centre = points.mean(axis=1)
+    r = points - centre[:, np.newaxis]  # centroid-to-vertex vectors
+    inertia_tensor = np.matmul(r, r.T)
+    u = np.linalg.svd(inertia_tensor)[0]
+    normal = matrix.col(u[:, 2].tolist()).normalize()
+    sum_vertices = matrix.col((0, 0, 0))
+    for v in vertices:
+        sum_vertices += v
+    if normal.dot(sum_vertices) > 0:
+        normal *= -1.0
+
+    # For multi-panel detectors cluster fast, slow axes by DBSCAN to get a
+    # consensus X, Y for the 2D plane
+    clustered_axes = False
+    if len(detector) > 1 and cluster:
+        clustered_axes = True
+        axes = []
+        for panel in detector:
+            axes.append(panel.get_fast_axis())
+            axes.append(panel.get_slow_axis())
+        clusters = cluster.DBSCAN(eps=0.1, min_samples=2).fit_predict(axes)
+        nclusters = max(clusters) + 1
+
+        # Revert to single panel mode if clustering is unsucessful
+        if nclusters < 2:
+            clustered_axes = False
+
+    if clustered_axes:
+        summed_axes = [matrix.col((0, 0, 0)) for i in range(nclusters)]
+        for axis, cluster in zip(axes, clusters):
```

Comment:
Definition of `cluster` here shadows the imported name, leading to the unexpected test failure